src/kakarot/accounts/library.cairo

```diff
@@ -15,64 +18,311 @@ from starkware.starknet.common.syscalls import (
+    // @notice This function is used to upgrade the smart contract account.
+    // @param new_class The new class of the account.
+    func upgrade{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(new_class: felt) {
+        // Access control check. Only the EOA owner should be able to upgrade its contract.
+        // with `supports_interface`
+        internal.assert_only_self();
+        // TODO: only valid classes should be allowed to be upgraded. Add a validation on the new class interface.
+        assert_not_zero(new_class);
+        replace_class(new_class);
+        Account_implementation.write(new_class);
+        return ();
+    }
+
+    @view
+    func get_implementation{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        implementation: felt
+    ) {
+        let (implementation) = Account_implementation.read();
+        return (implementation=implementation);
+    }
+
+    // @return address The EVM address of the account
     func get_evm_address{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
         address: felt
     ) {
-        let (address) = evm_address.read();
+        let (address) = Account_evm_address.read();
         return (address=address);
     }
 
+    // @notice This function checks if the account was initialized.
+    // @return is_initialized 1 if the account has been initialized 0 otherwise.
+    func is_initialized{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }() -> (is_initialized: felt) {
+        let is_initialized: felt = Account_is_initialized.read();
+        return (is_initialized=is_initialized);
+    }
+
+    // EOA functions
+
+    // Constants
+    // @dev see utils/interface_id.py
+    const INTERFACE_ID = 0x68bca1a;
+
+    struct Call {
+        to: felt,
+        selector: felt,
+        calldata_len: felt,
+        calldata: felt*,
+    }
+
+    // Tmp struct introduced while we wait for Cairo to support passing `[Call]` to __execute__
+    struct CallArray {
+        to: felt,
+        selector: felt,
+        data_offset: felt,
+        data_len: felt,
+    }
+
+    // @notice Validate the signature of every call in the call array.
+    // @dev Recursively validates if tx is signed and valid for each call -> see utils/eth_transaction.cairo
+    // @param call_array_len The length of the call array.
+    // @param call_array The call array.
+    // @param calldata_len The length of the calldata.
+    // @param calldata The calldata.
+    func validate{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        bitwise_ptr: BitwiseBuiltin*,
+        range_check_ptr,
+    }(call_array_len: felt, call_array: CallArray*, calldata_len: felt, calldata: felt*) -> () {
+        alloc_locals;
+        if (call_array_len == 0) {
```

Comment:
maybe add a comment to say that we do the check only at the end of the recursion